snap/pack/pack.go

```diff
@@ -107,27 +107,33 @@ func CheckSkeleton(w io.Writer, sourceDir string) error {
 }
 
 func loadAndValidate(sourceDir string) (*snap.Info, error) {
-	// ensure we have valid content
+	// Parsing of snap info is duplicated in ReadInfoFromSnapFile. It is done
+	// here to retrieve the snap instance name, if available, to use in case of
+	// ReadInfoFromSnapFile error
 	yaml, err := ioutil.ReadFile(filepath.Join(sourceDir, "meta", "snap.yaml"))
 	if err != nil {
 		return nil, err
 	}
-
 	info, err := snap.InfoFromSnapYaml(yaml)
 	if err != nil {
 		return nil, err
 	}
+	instanceName := info.InstanceName()
 
-	if err := snap.Validate(info); err != nil {
-		return nil, fmt.Errorf("cannot validate snap %q: %v", info.InstanceName(), err)
+	// ReadInfoFromSnapFile covers the following required steps:
+	// 	- Read snap metadata from meta/snap.yaml
+	//      - Parse snap info from meta/snap.yaml without side info
+	//      - Add and bind implicit hooks from meta/hooks
+	//      - Validate available snap information
+	//      - Validate snapshot metadata from opional meta/snapshot.yaml
+	info, err = snap.ReadInfoFromSnapFile(snapdir.New(sourceDir), nil)
+	if err != nil {
+		return nil, fmt.Errorf("cannot validate snap %q: %v", instanceName, err)
 	}
 
 	if err := snap.ValidateContainer(snapdir.New(sourceDir), info, logger.Noticef); err != nil {
```

Comment:
Noticed a bit of a chicken egg with this function after `ReadSnapshotYamlFromSnapFile` done inside `ReadInfoFromSnapFile`:

If e.g. `meta/snapshot.yaml` have incorrect permission, this will already cause an error during `ReadInfoFromSnapFile`.
On the other hand, `ValidateContainer` cannot be moved earlier due to dependency on Info.

Is this a deal breaker?

Reply:
it depends how common an issue we think this could be ?  if we don't want to use ReadInfoFromSnapFile then I think we need to keep more of the code here as is, but probably really try to share some code between here and ReadInfoFromSnapFile via some helpers in the internal package

Reply:
In retrospect I realised this situation is partly similar to that of meta/snap.yaml and arguable fine.
If the permission of meta/snap.yaml is incorrect (e.g. cannot read), then we also catch it implicitly with read failure and not with container validation. No functionality is lost, the failure just looks different for this specific case.

Reply:
I opted to revert back more to the way it was, preventing this issue.